src/cpu/testers/spatter_gen/spatter_gen.cc

```diff
@@ -0,0 +1,584 @@
+            requestBuffer.push(pkt);
+            requestBufferInsertionTime.push(curTick());
+            DPRINTF(
+                SpatterGen,
+                "%s: Pushed pkt: %s to requestBuffer. Access to VALUES.\n",
+                __func__, pkt->print()
+            );
+            // now deallocate resources for reading the index
+            delete idx_info;
+            delete idx_pkt;
+            int_reg_used_delta--;
+            receiveBuffer.pop();
+            receiveBufferInsertionTime.pop();
+        } else if (have_kernel && have_int_reg) {
+            SpatterKernel& front = kernels.front();
+            size_t index_size, value_size;
+            Addr index_addr, value_addr;
+            std::tie(index_size, index_addr) = front.nextIndexPair();
+            std::tie(value_size, value_addr) = front.nextValuePair();
+            // occupy an int register
+            available_int_regs_this_cycle--;
+            int_reg_used_delta++;
+            generatorBusyUntil[i] = clockEdge(Cycles(requestGenLatency));
+
+            AccessInfo* acc_info = new AccessInfo(
+                                                AccessType::INDEX,
+                                                front.type(),
+                                                value_size,
+                                                value_addr
+                                                );
+
+            RequestPtr req = std::make_shared<Request>(
+                                                    index_addr,
+                                                    index_size,
+                                                    0,
+                                                    requestorId
+                                                    );
+            // Dummy PC to have PC-based prefetchers latch on;
+            // get entropy into higher bits
+            req->setPC(((Addr)requestorId) << 2);
+            PacketPtr pkt = new Packet(req, MemCmd::ReadReq);
+            uint8_t* pkt_data = new uint8_t[req->getSize()];
+            // Randomly intialize pkt_data, for testing cache coherence.
+            for (int i = 0; i < req->getSize(); i++) {
+                pkt_data[i] = random_mt.random<uint8_t>();
+            }
+            pkt->dataDynamic(pkt_data);
+            pkt->pushSenderState(acc_info);
+
+            requestBuffer.push(pkt);
+            requestBufferInsertionTime.push(curTick());
+            DPRINTF(
+                SpatterGen,
+                "%s: Pushed pkt: %s to requestBuffer. Access to INDEX.\n",
+                __func__, pkt->print()
+            );
+
+            front.iterate();
+            if (front.done()) {
+                exitSimLoop(
+                    csprintf(
+                    "SpatterGen %s done generating "
+                    "requests from SpatterKernel %d.",
+                    name(), front.id()
+                    ),
+                    0,
+                    nextCycle()
+                );
```

Comment:
Why is this not done in `checkSimDone`?

Reply:
checkSimDone is for overall exit. This exit is per kernel. We're not completely done here. I think it is useful to try dump stats per kernel.